### ISSUE TITLE
No space left on device

### ISSUE THREAD

**Issue by wpfl-dbt** [i-dot-ai/redbox#179]
Core commands for running this, `docker compose up` and `make run`, will very quickly run into "no space left on device". I appreciate I can (and have) pruned, but docker has a pretty chunky 64Gb available to it. I suspect the memory footprint of this is much larger than you expect?

**Desktop:**
 - OS: macOS 14.4.1 (23E224)

**Comment by gecBurton** [comment]
have you pruned everything  https://github.com/i-dot-ai/redbox-copilot?tab=readme-ov-file#error-elasticsearch-137 8gb should be sufficient to run all the services

**Comment by gecBurton** [comment]
fair point re: telling you what you need! the core services (at the moment) are:
* core-api
* embedder
* ingester

and you will need some chat interface , which right now, is best served by streamlit-app

The above builds in CI on standard GH runners. I am afk at the moment but will be able to dig into your issue properly tomorrow.

**Comment by gecBurton** [comment]
yep. this is an excellent idea, please raise away!

**Comment by peterkwells** [comment]
hello, same issue here - temporarily resolved by expanding disk space available to Docker. In case it's helpful here's the amount of space being used by the various images.

<img width="1280" alt="Screenshot 2024-04-09 at 07 54 01" src="https://github.com/i-dot-ai/redbox-copilot/assets/6670237/89b17745-f73b-48bc-a517-0876a4a1d1e3">

On desktop, also running MacOS 14.4.1 (23E224)

**Comment by gecBurton** [comment]
This is not the answer, but it is a step in the right direction https://github.com/i-dot-ai/redbox-copilot/pull/200

**Comment by 1lliter8** [comment]
Some insights on this ticket after working in streamlit for a while:

* Lots of services use a builder/runtime from a common poetry lockfile. Can we use [docker extensions](https://docs.docker.com/compose/compose-file/11-extension/) to extend this pattern to the entire compose, then use [poetry sync](https://python-poetry.org/docs/managing-dependencies/#synchronizing-dependencies) to customise for each service. It might not save much space, but it'll definitely save time
* Check out `poetry show --tree` -- `sentence-transformers` has a dependency on PyTorch, and even though we're developing on CPU, it's installing 2-3Gb of Nvidia GPU dependencies. I switched the streamlit repo to CPU inference, but you have an issue that many of us are developing on Mac, but it runs on Debian, and PyTorch has different install processes for both which poetry doesn't like. I sorted by switching to dev containers, but it's a bit vendor lockin-y 
    * Seems [`sys-platform` markers are supported by poetry](https://stackoverflow.com/questions/64094979/how-can-i-specify-dependencies-on-operating-system-in-python-poetry) so maybe there's a non-dev-container route to this too
